### PR TITLE
Retry initiateMultipartUpload

### PR DESCRIPTION
**Pull request by sopel39**
```markdown
# Hive/Iceberg/Delta
* Improve resilience of insert queries. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectLineRecordReader.java [hunk 1]
```diff
@@ -212,6 +213,10 @@ private int readLine(Text value)
                     });
         }
         catch (Exception e) {
+            if (e instanceof InterruptedException) {
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectLineRecordReader.java:216]: > Reset interrupted flag


"Reset interrupted flag in S3 Select reader"

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
cc @alexjo2144 @findinpath @homar

**Review by homar — APPROVED** [review]

**Comment by alexjo2144** [comment]
Does this happen often? Curious what kind of errors from AWS are you seeing

**Comment by findinpath** [comment]
Could you please add some context / description to this fix?
It would be helpful to know in which situations does the initiation of the multipart upload request fail.

**Comment by sopel39** [comment]
The failure can happen during insert when there is high pressure on S3. Usually these are timeout exceptions

**Comment by findinpath** [comment]
> Usually these are timeout exceptions

Why don't we add a retry mechanism for the specific case of timeouts then?

**Comment by sopel39** [comment]
@findinpath this
```
                            if (e instanceof AmazonS3Exception) {
                                switch (((AmazonS3Exception) e).getStatusCode()) {
                                    case HTTP_FORBIDDEN, HTTP_NOT_FOUND, HTTP_BAD_REQUEST -> throw new UnrecoverableS3OperationException(bucket, key, e);
                                }
                            }
```
is what you will find in other places that retry S3 calls (in one form or other)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java [hunk 2]
```diff
@@ -1217,24 +1223,51 @@ private void customizePutObjectRequest(PutObjectRequest request)
...
-        InitiateMultipartUploadRequest request = new InitiateMultipartUploadRequest(bucket, key)
-                .withObjectMetadata(new ObjectMetadata())
-                .withCannedACL(s3AclType.getCannedACL())
-                .withRequesterPays(requesterPaysEnabled)
-                .withStorageClass(s3StorageClass.getS3StorageClass());
+        try {
+            return retry()
+                    .maxAttempts(maxAttempts)
+                    .exponentialBackoff(BACKOFF_MIN_SLEEP, maxBackoffTime, maxRetryTime, 2.0)
+                    .stopOn(InterruptedException.class, UnrecoverableS3OperationException.class)
```
Comment (pettyjamesm) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:1230]: This should probably stop on `AbortedException.class` as well, which is what the AWS SDK will actually throw if it catches an `InterruptedException` while executing the API request.  Similarly, you might want to reset `Thread.currentThread().interrupt()` for that case too.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java [hunk 3]
```diff
@@ -1217,24 +1223,51 @@ private void customizePutObjectRequest(PutObjectRequest request)
...
+                                switch (((AmazonS3Exception) e).getStatusCode()) {
+                                    case HTTP_FORBIDDEN, HTTP_NOT_FOUND, HTTP_BAD_REQUEST -> throw new UnrecoverableS3OperationException(bucket, key, e);
+                                }
+                            }
+                            throw e;
+                        }
+                    });
+        }
+        catch (Exception e) {
+            if (e instanceof InterruptedException) {
```
Comment (pettyjamesm) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:1265]: `if (e instanceof InterruptedException || e instanceof AbortedException) {`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectLineRecordReader.java (hunk 1 above)
Comment (pettyjamesm) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3select/S3SelectLineRecordReader.java:216]: Same thing here about the SDK client throwing `AbortedException` when it encounters an `InterruptedException` internally.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java (hunk 2 above)
Comment (sopel39) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:1230]: Follow up https://github.com/trinodb/trino/pull/15578

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java (hunk 3 above)
Comment (sopel39) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java:1265]: Follow up https://github.com/trinodb/trino/pull/15578

**Pull request merged**